pkg/graveler/committed/diff.go

```diff
@@ -112,61 +174,95 @@ func (d *diffIterator) Next() bool {
-			return true
-		case diffItCompareResultRightBeforeLeft:
-			// nothing on left, or right before left
-			d.currentDiff = &graveler.Diff{Type: graveler.DiffTypeAdded, Key: d.rightValue.record.Key.Copy(), Value: d.rightValue.record.Value}
 			d.rightValue.record, d.rightValue.rng, d.rightValue.err = diffIteratorNextValue(d.right)
-			return true
 		}
+		for {
+			if d.rightValue.err != nil {
+				d.err = d.rightValue.err
+			}
+			if d.leftValue.err != nil {
+				d.err = d.leftValue.err
+			}
+			if d.err != nil {
+				d.currentDiff = nil
+				d.clearCurrentRange()
+				return false
+			}
+			compareResult := d.compareDiffIterators()
+			switch compareResult {
+			case diffItCompareResultDone:
+				d.currentDiff = nil
+				return false
+			case diffItCompareResultSameRanges:
+				d.leftValue.record, d.leftValue.rng, d.leftValue.err = diffIteratorNextRange(d.left)
+				d.rightValue.record, d.rightValue.rng, d.rightValue.err = diffIteratorNextRange(d.right)
+			case diffItCompareResultLeftRangeBeforeRight:
+				d.setCurrentRangeLeft()
+				return true
+			case diffItCompareResultRightRangeBeforeLeft:
+				d.setCurrentRangeRight()
+				return true
+			case diffItCompareResultSameKeys:
+				// same keys on different ranges
+				d.currentDiff = &graveler.Diff{Type: graveler.DiffTypeChanged, Key: d.rightValue.record.Key.Copy(), Value: d.rightValue.record.Value, LeftIdentity: d.leftValue.record.Identity}
+				d.leftValue.record, d.leftValue.rng, d.leftValue.err = diffIteratorNextValue(d.left)
+				d.rightValue.record, d.rightValue.rng, d.rightValue.err = diffIteratorNextValue(d.right)
+				return true
+			case diffItCompareResultSameIdentities, diffItCompareResultNeedStartRangeBoth:
+				d.leftValue.record, d.leftValue.rng, d.leftValue.err = diffIteratorNextValue(d.left)
+				d.rightValue.record, d.rightValue.rng, d.rightValue.err = diffIteratorNextValue(d.right)
+			case diffItCompareResultNeedStartRangeLeft:
+				d.leftValue.record, d.leftValue.rng, d.leftValue.err = diffIteratorNextValue(d.left)
+			case diffItCompareResultNeedStartRangeRight:
+				d.rightValue.record, d.rightValue.rng, d.rightValue.err = diffIteratorNextValue(d.right)
+			case diffItCompareResultLeftBeforeRight:
+				// nothing on right, or left before right
+				d.currentDiff = &graveler.Diff{Type: graveler.DiffTypeRemoved, Key: d.leftValue.record.Key.Copy(), Value: d.leftValue.record.Value, LeftIdentity: d.leftValue.record.Identity}
+				d.leftValue.record, d.leftValue.rng, d.leftValue.err = diffIteratorNextValue(d.left)
+				return true
+			case diffItCompareResultRightBeforeLeft:
+				// nothing on left, or right before left
+				d.currentDiff = &graveler.Diff{Type: graveler.DiffTypeAdded, Key: d.rightValue.record.Key.Copy(), Value: d.rightValue.record.Value}
+				d.rightValue.record, d.rightValue.rng, d.rightValue.err = diffIteratorNextValue(d.right)
+				return true
+			}
+		}
+	}
+}
+
+func (d *diffIterator) NextRange() bool {
```

Comment:
Given the comment on `ErrNoRange`, I believe that now _this_ function needs to be documented.  Specifically, AFAICT it cannot be called twice in a row (unlike a regular iterator).  That's OK, it's just different -- and requires documentation.